### PR TITLE
#52 return 402 when wallet is pushed with negative balance

### LINKED ISSUE #52: 402 on PUT is not returned when taxes are not paid
Let's return 402 (taxes not paid) on PUT if the wallet has a big debt.

### PR DESCRIPTION
**Pull request by jedediah**
Fixes #52

- Add namespace for custom exceptions `Zold::Error::`
- Node returns 402 Payment Required from negative balance wallet push

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#376](https://github.com/zold-io/zold/issues/376) is now in scope, role is `REV`

File: lib/zold/error.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# SOFTWARE.
+
+module Zold
+  # Errors raised from engine classes
+  module Error
+    # Wallet failed validation
+    class InvalidWallet < StandardError
+    end
+
+    # Wallet is in debt
```
Comment (ledestin) [inline lib/zold/error.rb]: @jedediah any reason not to use the comment as the class name?

File: lib/zold/node/safe_entrance.rb [hunk 2]
```diff
@@ -60,16 +61,19 @@ def push(id, body)
...
         if balance.negative? && !wallet.root?
-          raise "The balance #{balance} of #{wallet.id} is negative and it's not a root wallet"
+          raise Error::NegativeBalance,
+                "The balance #{balance} of #{wallet.id} is negative and it's not a root wallet"
         end
         Emission.new(wallet).check
         tax = Tax.new(wallet)
         if tax.in_debt?
-          raise "Taxes are not paid, can't accept the wallet; the debt is #{tax.debt} (#{tax.debt.to_i} zents)"
+          raise Error::NegativeBalance,
```
Comment (ledestin) [inline lib/zold/node/safe_entrance.rb]: @jedediah it looks like you're using the same exception for different errors. As far as I know, tax debt doesn't automatically mean negative balance.

File: test/node/test_front.rb [hunk 3]
```diff
@@ -113,6 +113,19 @@ def test_pushes_twice
     end
   end
 
+  def test_402_on_negative_balance
+    FakeHome.new.run do |home|
```
Comment (ledestin) [inline test/node/test_front.rb]: @jedediah the same question - does tax debt always mean negative balance?

**Comment by ledestin** [comment]
@jedediah please see my comments

**Comment by 0crat** [comment]
@ledestin[/z](https://www.0crat.com/u/ledestin) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by ledestin** [comment]
@jedediah ping

File: test/node/test_front.rb (hunk 3 above)
Comment (jedediah) [reply test/node/test_front.rb]: This test only covers negative balance, not tax debt.

Comment (ledestin) [reply test/node/test_front.rb]: @jedediah I'm confused, the issue is about tax debt, but you focus on negative balance?

**Comment by yegor256** [comment]
@jedediah this is already fixed, thanks

**Comment by ypshenychka** [comment]
@jedediah According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>Messages in a job [always](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html#4-avoid-noiseaddress-your-comments) start with a name of a user they are addressed to.

Please correct your code review [comment](https://github.com/zold-io/zold/pull/376#discussion_r203512718) by indicating an addressee in the beginning.

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @ledestin[/z](https://www.0crat.com/u/ledestin)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request closed without merging**